File: hphp/hack/src/typing/typing_check_utils.ml
(*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Hh_prelude

let type_file ctx fn ~full_ast =
  let (funs, classes, typedefs, consts, modules) = Nast.get_defs full_ast in
  let (errors, tast) =
    Errors.do_with_context fn Errors.Typing (fun () ->
        let check ~f defs =
          List.map defs ~f:snd
          |> List.filter_map ~f:(fun full_ast -> f ctx ~full_ast)
        in
        let fs = check ~f:Typing_check_job.type_fun funs |> List.concat in
        let cs = check ~f:Typing_check_job.type_class classes in
        let ts = check ~f:Typing_check_job.check_typedef typedefs in
        let gcs = check ~f:Typing_check_job.check_const consts in
        let mds = check ~f:Typing_check_job.check_module modules in
        fs @ cs @ ts @ gcs @ mds)
  in
  (tast, errors)

File: hphp/hack/src/hackrs/rupro/typing_check_utils.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use std::rc::Rc;

use oxidized::aast;
use oxidized::nast;
use ty::reason::Reason;

use crate::errors::HackError;
use crate::parsing_error::ParsingError;
use crate::tast;
use crate::typing::typing_error::Result;
use crate::typing_check_job::TypingCheckJob;
use crate::typing_ctx::TypingCtx;

pub struct TypingCheckUtils;

impl TypingCheckUtils {
    pub fn type_file<R: Reason>(
        ctx: Rc<TypingCtx<R>>,
        ast: &nast::Program,
        parsing_errors: Vec<ParsingError>,
    ) -> Result<(tast::Program<R>, Vec<HackError<R>>)> {
        use aast::Def;
        let mut tast = vec![];
        let mut errs = vec![];
        errs.extend(parsing_errors.into_iter().map(Into::into));
        let mut accumulate = |(new_def, new_errs)| {
            tast.push(new_def);
            errs.extend(new_errs);
        };
        for def in ast.defs() {
            let ctx = Rc::clone(&ctx);
            match def {
                Def::Fun(f) => accumulate(TypingCheckJob::type_fun(ctx, &*f)?),
                Def::Class(c) => accumulate(TypingCheckJob::type_class(ctx, &*c)?),
                Def::Typedef(t) => accumulate(TypingCheckJob::type_typedef(ctx, &*t)?),
                Def::Constant(cst) => accumulate(TypingCheckJob::type_const(ctx, &*cst)?),
                Def::Namespace(_)
                | Def::Stmt(_)
                | Def::Module(_)
                | Def::NamespaceUse(_)
                | Def::SetNamespaceEnv(_)
                | Def::SetModule(_)
                | Def::FileAttributes(_) => unreachable!("Program::defs won't emit these"),
            }
        }
        Ok((aast::Program(tast), errs))
    }
}
